### ISSUE TITLE
Colab Memory crash while training eager_few_shot_od_training_tflite

### ISSUE THREAD

**Issue by Jahanzeb-jogo** [tensorflow/models#9883]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.
- I am using few shot object detection training with tensorflow 2.2 in google colab and I have tried tf-nighty too

## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/blob/master/research/object_detection/colab_tutorials/eager_few_shot_od_training_tf2_colab.ipynb

## 2. Describe the bug

I am running your eager_few_shot_od_training_tflite through colab pro with some custom inference code for 2 classes, So while training the model for 2 epoch or more than 2, the colab memory Ram crashes as it becomes full with no log for the error and I have to run the whole process again, I wanted to know how can I overcome this issue, Plus How much dataset size I can use using this notebook?

These are the parameter I am using
num_examples = 25
batch_size = 8
learning_rate = 0.001
num_batches = 100
num_epochs = 3

## 3. My code:

tf.keras.backend.set_learning_phase(True)

num_examples = 25
batch_size = 8
learning_rate = 0.001
num_batches = 100
num_epochs = 3

trainable_variables = detection_model.trainable_variables
to_fine_tune = []
prefixes_to_train = [
  'WeightSharedConvolutionalBoxPredictor/WeightSharedConvolutionalBoxHead',
  'WeightSharedConvolutionalBoxPredictor/WeightSharedConvolutionalClassHead']
for var in trainable_variables:
  if any([var.name.startswith(prefix) for prefix in prefixes_to_train]):
    to_fine_tune.append(var)

def get_model_train_step_function(model, optimizer, vars_to_fine_tune):

  @tf.function
  def train_step_fn(image_tensors,
                    groundtruth_boxes_list,
                    groundtruth_classes_list):
  
    shapes = tf.constant(batch_size * [[MODEL_INPUT_SIZE, MODEL_INPUT_SIZE, 3]], dtype=tf.int32)
    model.provide_groundtruth(
        groundtruth_boxes_list=groundtruth_boxes_list,
        groundtruth_classes_list=groundtruth_classes_list)
    
    with tf.GradientTape() as tape:
      
      preprocessed_images = tf.concat([detection_model.preprocess(image_tensor)[0] for image_tensor in image_tensors], axis=0)

    
      prediction_dict = model.predict(preprocessed_images, shapes)

   
      losses_dict = model.loss(prediction_dict, shapes)

    
      total_loss = losses_dict['Loss/localization_loss'] + losses_dict['Loss/classification_loss']

    
      gradients = tape.gradient(total_loss, vars_to_fine_tune)
      optimizer.apply_gradients(zip(gradients, vars_to_fine_tune))

    return total_loss

  return train_step_fn

optimizer = tf.keras.optimizers.Adam(learning_rate=learning_rate)
train_step_fn = get_model_train_step_function(detection_model, optimizer, to_fine_tune)

print('Start fine-tuning!', flush=True)

train_cnt = 0
val_cnt = 0
for e in range(num_epochs):

  print(f"Epoch: {e}")

  for idx in range(num_batches):
   
    all_keys = list(range(len(train_images_np)))
    random.shuffle(all_keys)
    example_keys = all_keys[:batch_size]

    gt_boxes_list = [gt_box_tensors[key] for key in example_keys]
    gt_classes_list = [gt_classes_one_hot_tensors[key] for key in example_keys]
    image_tensors = [train_image_tensors[key] for key in example_keys]
   
    total_loss = train_step_fn(image_tensors, gt_boxes_list, gt_classes_list)

    if idx % 10 == 0:
      print('batch ' + str(idx) + ' of ' + str(num_batches) + ', loss=' +  str(total_loss.numpy()), flush=True)

   
      log_value("train_loss", float(total_loss.numpy()), train_cnt * 10)
   
      writer.flush()
      train_cnt = train_cnt + 1

      val_loss = train_step_fn(val_image_tensors, val_gt_box_tensors, val_gt_classes_one_hot_tensors)
      print("Val: ", val_loss.numpy())
      log_value("val_loss", float(val_loss.numpy()), val_cnt * 10)
      val_cnt = val_cnt + 1
      wandb.log({"train_loss": total_loss.numpy(),
              "val_loss": val_loss.numpy(),
              "BATCH": idx,
             "learning_rate": learning_rate
   })

print('Done fine-tuning!')
ckpt_manager.save()
print('Checkpoint saved!')

## 4 Steps to reproduce

Tried tensorflow 2.2 and other tf-nighty versions

**Comment by Jahanzeb-jogo** [comment]
@srjoglekar246

**Comment by srjoglekar246** [comment]
Hey @pkulzc could you PTAL?

@Jahanzeb-jogo Coud you check the code formatting to make it a bit easier to read? :-) Also, I don't see what error you get? Can you copy-paste the logs if any?

**Comment by jvishnuvardhan** [comment]
@Jahanzeb-jogo Is this still an issue for you? I ran the colab with custom dataset (2 classes- similar to you) and I didn't notice any colab crash. 

Can you please check with recent TF versions and let us know whether it is still an issue for you. Thanks!
